src/placeholder/parameter/parameter.py

```diff
@@ -0,0 +1,298 @@
+    ----------
+    All attributes from :class:`~placeholder.parameter.parameter.Variable`s in `variables` 
+    are carried over but are put into a list.
+
+    num_fe: int
+        total number of effects (variables) for the parameter
+
+    """
+    param_name: str
+    link_fun: Callable
+    variables: InitVar[List[Variable]]
+
+    num_fe: int = field(init=False)
+    covariate: List[str] = field(init=False)
+    var_link_fun: List[Callable] = field(init=False)
+
+    fe_init: List[float] = field(init=False)
+    re_init: List[float] = field(init=False)
+
+    fe_gprior: List[List[Prior]] = field(init=False)
+    re_gprior: List[List[Prior]] = field(init=False)
+    fe_bounds: List[List[Prior]] = field(init=False)
+    re_bounds: List[List[Prior]] = field(init=False)
+
+    re_zero_sum_std: List[float] = field(init=False)
+
+    def __post_init__(self, variables):
+        assert isinstance(variables, list)
+        assert len(variables) > 0
+        assert isinstance(variables[0], Variable)
+        self.num_fe = len(variables)
+        for k, v in consolidate(Variable, variables).items():
+            self.__setattr__(k, v)
+
+
+@dataclass
+class ParameterFunction:
+    """A class for function on parameters.
+
+    Parameters
+    ----------
+    param_function_name: str
+        name of the parameter function
+    param_function: callable
+        parameter function
+    param_function_fe_gprior: List[float]
+        a list of two floats specifying mean and std for Gaussian prior on the function.
+    """
+
+    param_function_name: str
+    param_function: Callable
+    param_function_fe_gprior: List[Prior] = field(default_factory=lambda: [Prior])
+
+    def __post_init__(self):
+        assert isinstance(self.param_function_name, str)
+
+
+@dataclass
+class ParameterSet:
+    """A class for a set of parameters.
+
+    Parameters
+    ----------
+    parameters: List[:class:`~placeholder.parameter.parameter.Parameter`]
+        a list of parameters.
+    parameter_functions: List[:class:`~placeholder.parameter.parameter.ParameterFunction`]
+        a list of parameter functions.
+
+    Attributes
+    ----------
+    All attributes from :class:`~placeholder.parameter.parameter.Parameter`s in `parameters` 
+    are carried over and put into a list of lists.
+
+    num_fe: int
+        total number of effects (variables) for the parameter set.
+    """
+
+    parameters: InitVar[List[Parameter]]
+    parameter_functions: InitVar[List[ParameterFunction]] = None
+
+    param_name: List[str] = field(init=False)
+    num_fe: int = field(init=False)
+    link_fun: List[Callable] = field(init=False)
+    covariate: List[List[str]] = field(init=False)
+    var_link_fun: List[List[Callable]] = field(init=False)
+
+    fe_init: List[List[float]] = field(init=False)
+    re_init: List[List[float]] = field(init=False)
+
+    fe_gprior: List[List[List[Prior]]] = field(init=False)
+    re_gprior: List[List[List[Prior]]] = field(init=False)
```

Comment:
I am having trouble understanding how the actual mixed models (where you estimate both fixed and variance parameters, sometimes both for random effects and groups' observation errors) could fit into this framework. 

It looks like this notion of ParameterSet explicitly assumes that:
1) Variance parameters are known and given through Prior 
2) The inference is happening for the concatenated vector of fixed and random effects, but not for the fixed effects and _variances_ of random effects.
It looks a bit limiting if it does not allow any variance inference.

Reply:
For `Prior` I am thinking that the distributions are all known. For something that we actually want to estimate, that's a new ParameterSet, and we also need a Likelihood class. I'll put something together before our meeting to explain.

Reply:
My guess is that I define variance parameters as fixed parameters with no random effects, right?

Reply:
You define variance parameters as ParameterSet, but they could have random effects if you wanted to

Reply:
Well, random effects for _variance_ parameters sounds like an overkill with some second-order uncertainty :) Never seen this.

Reply:
haha well then you can not do that instead (but DisMod does have this XD)

Reply:
Okay, I will rephrase it like this:
Imagine you have a set of fixed parameters \beta and respective random effects for beta: u_i. Every u_i ~ N(0, \gamma), where \gamma is unknown. You want to do inference for (\beta, \gamma). 
Effectively, \gamma is a Gaussian prior for \beta but you can't define it in advance. You also can't treat them as two independent sets of parameters because they are actually related.
My question is: what would be the correct way to describe this construction in terms of our framework?

Reply:
I'm trying to write this out right now

Reply:
Lol, sorry, I was commenting from the code page, but when I reloaded it half of you comments jumped in between mine :D Did not see them before writing. 
Sure, we can discuss it later, no rush!